govcd/nsxv_firewall.go

```diff
@@ -0,0 +1,225 @@
+	firewallRuleRequest := requestEdgeFirewallRules{
+		EdgeFirewallRules: []*types.EdgeFirewallRule{firewallRuleConfig},
+	}
+
+	httpPath, err := egw.buildProxiedEdgeEndpointURL(types.EdgeCreateFirewallPath)
+	if err != nil {
+		return nil, fmt.Errorf("could not get Edge Gateway API endpoint: %s", err)
+	}
+	// We expect to get http.StatusCreated or if not an error of type types.NSXError
+	// The query must be wrapped differently, depending if it mus specify the "aboveRuleId" parameter
+	var resp *http.Response
+	if aboveRuleId == "" {
+		resp, err = egw.client.ExecuteRequestWithCustomError(httpPath, http.MethodPost, types.AnyXMLMime,
+			"error creating firewall rule: %s", firewallRuleRequest, &types.NSXError{})
+	} else {
+		errString := fmt.Sprintf("error creating firewall rule (aboveRuleId: %s): %%s", aboveRuleId)
+		resp, err = egw.client.ExecuteParamRequestWithCustomError(httpPath, params, http.MethodPost, types.AnyXMLMime,
+			errString, firewallRuleConfig, &types.NSXError{})
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	// Location header should look similar to:
+	// [/network/edges/edge-1/firewall/config/rules/197157]
+	firewallRuleId, err := extractNsxObjectIdFromPath(resp.Header.Get("Location"))
+	if err != nil {
+		return nil, err
+	}
+
+	readFirewallRule, err := egw.GetNsxvFirewallRuleById(firewallRuleId)
+	if err != nil {
+		return nil, fmt.Errorf("unable to retrieve firewall rule with ID (%s) after creation: %s",
+			firewallRuleId, err)
+	}
+	return readFirewallRule, nil
+}
+
+// UpdateNsxvFirewallRule updates types.EdgeFirewallRule with all fields using proxied NSX-V API. ID is
+// mandatory to perform the update.
+func (egw *EdgeGateway) UpdateNsxvFirewallRule(firewallRuleConfig *types.EdgeFirewallRule) (*types.EdgeFirewallRule, error) {
+	err := validateUpdateNsxvFirewallRule(firewallRuleConfig, egw)
+	if err != nil {
+		return nil, err
+	}
+
+	httpPath, err := egw.buildProxiedEdgeEndpointURL(types.EdgeCreateFirewallPath + "/" + firewallRuleConfig.ID)
+	if err != nil {
+		return nil, fmt.Errorf("could not get Edge Gateway API endpoint: %s", err)
+	}
+
+	// Result is either 204 for success, or an error of type types.NSXError
+	_, err = egw.client.ExecuteRequestWithCustomError(httpPath, http.MethodPut, types.AnyXMLMime,
+		"error while updating firewall rule : %s", firewallRuleConfig, &types.NSXError{})
+	if err != nil {
+		return nil, err
+	}
+
+	readFirewallRule, err := egw.GetNsxvFirewallRuleById(firewallRuleConfig.ID)
+	if err != nil {
+		return nil, fmt.Errorf("unable to retrieve firewall rule with ID (%s) after update: %s",
+			readFirewallRule.ID, err)
+	}
+	return readFirewallRule, nil
+}
+
+// GetNsxvFirewallRuleById retrieves types.EdgeFirewallRule by firewall rule ID as shown in the UI using proxied
+// NSX-V API.
+// It returns and error `ErrorEntityNotFound` if the firewall rule is not found
+func (egw *EdgeGateway) GetNsxvFirewallRuleById(id string) (*types.EdgeFirewallRule, error) {
```

Comment:
Get and delete use internal IDs or UI IDs. From comment I am not sure

Reply:
Thanks. Improved comments.